### ISSUE TITLE
Run tests in the order they are given on the command line

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#5520]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#5520|
|Type|enhancement|
|Created|2012-03-10 01:04:07Z|
|Branch|https://github.com/twisted/twisted/tree/trial-stable-argument-order-5520-3|

Sometimes the order of unit tests is important.  In these unfortunate cases, it's nice to be able to isolate the buggy code by reducing the number of tests running until a minimal set that reproduces the problem is found.  However, it's very difficult to do this with trial because trial essentially randomizes the order of the tests it runs.

  $ trial foo bar

This should run all tests discovered under `foo` and then all tests discovered under `bar`.

Attachments:

* [ordered-tests.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b8/0b8d6a6acdc63f8cdc3381b974993b52793b3ede/c406cb40c6b1ac02be57ced3f3e8c8299b23ac89.patch) (3476 bytes) - added by Julian on 2012-07-24 00:37:00Z - 
* [ordered-tests-2.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b8/0b8d6a6acdc63f8cdc3381b974993b52793b3ede/ef9beb326983b89b4cd5469f9d95621207e72778.patch) (3298 bytes) - added by Julian on 2012-08-07 13:24:02Z - 
* [ordered-tests-loadByNames.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b8/0b8d6a6acdc63f8cdc3381b974993b52793b3ede/54d1ae1442d06aafd00ff3bada6e62577aa51154.patch) (2056 bytes) - added by Julian on 2012-08-20 23:54:51Z - 
* [orderedTests-3.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b8/0b8d6a6acdc63f8cdc3381b974993b52793b3ede/f5205c5da8335335bf622ecd83b718f0472be368.patch) (2510 bytes) - added by Julian on 2012-09-02 02:43:26Z - 
* [ordered-tests-4.patch](https://raw.githubusercontent.com/twisted/twistedmatrix.com-trac-attachments/trunk/ticket/0b8/0b8d6a6acdc63f8cdc3381b974993b52793b3ede/449f6e8a938be6b2f6a443bb783ec24db8c46c20.patch) (1888 bytes) - added by Julian on 2013-01-23 01:13:36Z - 

<details><summary>Searchable metadata</summary>

```
trac-id__5520 5520
type__enhancement enhancement
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_trial_stable_argument_order_5520_3 branches/trial-stable-argument-order-5520-3
branch_author__Julian__exarkun__glyph Julian, exarkun, glyph
status__closed closed
resolution__fixed fixed
component__trial trial
keywords__ 
time__1331341447000000 1331341447000000
changetime__1362360503000000 1362360503000000
version__None None
owner__exarkun exarkun
cc__jml cc__cyli cc__acapnotic cc__namn
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="cyli's avatar" src="https://avatars.githubusercontent.com/u/223200?s=50" width="50" height="50">](https://github.com/cyli)<a name="note_2"></a>|@cyli commented|
|-|-|

This bug is related to (if slightly different than) [#1948](https://github.com/twisted/twisted/issues/1948) - many other tickets have been closed as duplicates to that one.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_5"></a>|@Julian commented|
|-|-|

First attempt in ordered-tests.patch.

If trial.Options()['tests'] happens to not be used in third party code, it's not used elsewhere outside trial's script, so we could outright just use an OrderedDict if there's any resistance towards adding the class added in the patch.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_6"></a>|@exarkun set owner to @Julian|
|-|-|

Thanks a lot for tackling this.

  1. Python already has a mutable ordered container - `list`.   I see the attraction of the `OrderedSet` approach, in that it avoids requiring changes to to the `Options['tests']` interface (and all relying code), but it's probably a short term benefit and a long term maintenance burden.  Apart from that, if we stick with this approach, the implementation needs documentation and direct unit tests.  And there's not much relying code (and the transformation is easy, from `add` to `append` and `update` to `extend`).  The only other question to consider is whether we _want_ the deduplicating behavior of a set as compared to a list.  I suggest that `trial foo foo` running `foo` twice is another feature we want.
  1. Thanks for updating the man page!  Most people ignore those. :)
  1. Please make the test method docstring more affirmative.  Avoid using the word "should" and just say that it does.
  1. Please use _camelCase_ instead of _under_scores_.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_7"></a>|@Julian removed owner|
|-|-|

Hey. Thanks for the review. Addressed your points I hope.

For 1, as I mentioned, I kept a set-like interface not for convenience just because I was unsure whether there was anything written on top of `trial` that relied on it being a set (`Options['tests']`) seemed public-ky, so, good to know (or maybe not so good :) that that's not the case.

**Comment by twisted-trac** [comment]
|<img alt="jaredsohn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_8"></a>|jaredsohn commented|
|-|-|

While reviewing this, it seems like if you make this change you will need to follow the CompatibilityPolicy (http://twistedmatrix.com/trac/wiki/CompatibilityPolicy) for this change in public interface.  You could perhaps deprecate 'tests' and create a new member, or just deprecate 'tests' and use a private '_tests' if nobody is using it or will use it in the future.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_9"></a>|@Julian commented|
|-|-|

Hey. Thanks. The original patch / implementation took that into account (and didn't change the interface of `tests`), but it sounded like exarkun was OK with changing it. Perhaps I should ask for clarification then from him.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_10"></a>|@Julian commented|
|-|-|

As per a verification on IRC:

20:02:20 < tos9> exarkun: Hey. If you happen to get a chance at any point, I know things are busy, can you comment on an issue raised in 5520 on whether a change is public API 
                 or not? The only reason I ask is it's something we briefly discussed before.
20:07:27 < exarkun> tos9: I'm ready to override the compatibility policy in this circumstance.  This is a highly obscure "API" with no direct documentation or tests, and 
                    changing from a set to a list is almost compatible anyway.
20:08:23 < exarkun> tos9: Strictly speaking, it is an incompatible change, and if we wanted to be completely stringent about it, we would keep the existing API exactly and only 
                    add the new API we need (and deprecate the old one).
20:10:42 < tos9> exarkun: Right, that's what I was thinking too. Can I quote you on the ticket?
20:11:35 < exarkun> tos9: Please do. 
20:11:39 < tos9> Great. Appreciated.

(Thanks for bringing it up, it was worth clarifying).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_11"></a>|@exarkun set owner to @exarkun|
|-|-|

Thanks.  I'm happy with this patch now.  I'll see if buildbot agrees and apply it if so.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_12"></a>|@exarkun commented|
|-|-|

(In [35310]) Branching to 'trial-stable-argument-order-5520'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_13"></a>|@exarkun commented|
|-|-|

(In [35311]) Apply patch with conflict resolved

refs [#5520](https://github.com/twisted/twisted/issues/5520)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_14"></a>|@exarkun set owner to @Julian|
|-|-|

Hmm.  I looked at the patch and thought really hard, and thought it would be good.  But then I actually tried using it, and I don't observe the desired behavior.  For example:

```

exarkun@top:~/Projects/Twisted/trunk$ bin/trial twisted.trial.test.test_warning.CollectWarningsTests.test_callsObserver twisted.trial.test.test_tests.TestUnhandledDeferred.test_isReported
twisted.trial.test.test_tests
  TestUnhandledDeferred
    test_isReported ...                                                    [OK]
twisted.trial.test.test_warning
  CollectWarningsTests
    test_callsObserver ...                                                 [OK]

-------------------------------------------------------------------------------
Ran 2 tests in 0.090s
```

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_15"></a>|@Julian removed owner|
|-|-|

Hi. Sorry for the trouble.

Couldn't reproduce with the example you posted, but after trying for a tiny bit I found another example, which suggested I missed a thing that was jiggling the order but only sometimes due to hash quirks. I fixed it I hope.

Attached an incremental fix for it, with a test case (I'm not sure what I did is the best way to test this, but I couldn't think of a better way. What I did was just call the relevant function twice with two differently ordered lists. I would assume that this should produce two consistently different internally ordered sets [of course I've run it and checked to see it passes after the fix on my machine] but if you have a better way please let me know).

Cheers.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_16"></a>|@Julian commented|
|-|-|

OK, I've attempted to offset possible hash-related uncertainty by fixing the tests to add a few more cases.

Think this is OK now, for any reviewer the latest file added (orderedTests-3.patch) applied onto the current branch's tip should be what I'm submitting here.

Thanks in advance.

**Comment by twisted-trac** [comment]
|<img alt="namn's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50"><a name="note_17"></a>|namn commented|
|-|-|

Shouldn't the last square bracket on line 490 in ```test_loader.py``` be moved to its own line?

And, line 492 could use some generator expression instead of ```map``` because we do not need a list and throw it away right after that.

Looks good to me otherwise. Thanks.

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_18"></a>|@Julian commented|
|-|-|

Hi.

Your comments got lost in my inbox, sorry 'bout that. If you can, next time remove the `review` keyword so it's a bit more obvious that there's something for me to address.

As for addressing them, I personally do put the closing bracket on the next line, but Twisted seems to like it on the same line, so I copied that. And the `map` is basically just copied from the other tests in the area as well. I have no idea if `TestSuite` can handle arbitrary iterators but I don't think there's any harm in keeping it as is.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_19"></a>|@glyph commented|
|-|-|

(In [36528]) Apply `orderedTests-3.patch` since that seems to be the most recent submission.

refs [#5520](https://github.com/twisted/twisted/issues/5520)

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_20"></a>|@glyph commented|
|-|-|

(In [36529]) Branching to 'trial-stable-argument-order-5520-2'

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_21"></a>|@glyph commented|
|-|-|

(In [36530]) Merge forward, resolve conflicts.

refs [#5520](https://github.com/twisted/twisted/issues/5520)

**Comment by twisted-trac** [comment]
|[<img alt="tomprince's avatar" src="https://avatars.githubusercontent.com/u/283816?s=50" width="50" height="50">](https://github.com/tomprince)<a name="note_22"></a>|@tomprince commented|
|-|-|

[buildbot run](https://buildbot.twistedmatrix.com/builders/documentation/builds/2798)

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun set owner to @exarkun<br>@exarkun set status to `assigned`|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_24"></a>|@exarkun set owner to @Julian<br>@exarkun set status to `new`|
|-|-|

Thanks.

  1. The `test_script.py` change introduces a dependency going from trial's test suite onto Conch's test suite.  This isn't an allowed dependency.  More generally, the change introduces a dependency from trial's test suite onto distant, somewhat arbitrary "application" code.  Changing Conch's test suite shouldn't break trial's test suite, but it might break this test if it happens to move or rename the Conch test this test now identifies.  So, instead, use some test names from the dummy test suite already in trial's test suite - eg `twisted.trial.test.mockcustomsuite`.
  1. `TestArgumentOrderTests` should have a docstring
  1. `test_preserveArgumentOrder` has an extra blank line at the end of its docstring and after its docstring.
  1. I think this is probably a feature, not a bugfix.  It's not a violation of any particular specification or a mishandled corner case of some established (tested/documented) feature, just a behavior that we want that trial didn't previously provide.

These all seem minor, I'll try to re-review quickly once they're addressed (feel free to re-assign to me for review if you like).

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_25"></a>|@Julian set owner to @exarkun|
|-|-|

Hey thanks for catching all those.

Patch to address.

mockcustomsuite and friends couldn't be used since they define empty test_suites, so I used other trial modules. Hope that's OK.

Re #4: the topfile was also in the wrong place entirely, it was at 5220 rather than 5520. That was confusing. Fix that dyxlesia glyph. Oh and, PEP 257 recommends docstrings with the extra lines (though I keep trying to not put them since I see Twisted doesn't use them). This one slipped through accidentally :).

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_26"></a>|@exarkun set status to `assigned`|
|-|-|

> Oh and, PEP 257 recommends docstrings with the extra lines

Ah right.  That's an outdated recommendation now anyway, as recent versions of python-mode in emacs can reflow the text without messing up the closing triple quote.

Also, you did it again in the new class docstring you added with this patch. ;)  I'll correct that and then merge.  Thanks!

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_27"></a>|@exarkun commented|
|-|-|

(In [36897]) Branching to 'trial-stable-argument-order-5520-3'

**Comment by twisted-trac** [comment]
|[<img alt="Julian's avatar" src="https://avatars.githubusercontent.com/u/329822?s=50" width="50" height="50">](https://github.com/Julian)<a name="note_28"></a>|@Julian commented|
|-|-|

Interesting. Didn't know that.

Also, gah. Sorry. Thank you.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_29"></a>|@exarkun set status to `closed`|
|-|-|

(In [36899]) Merge trial-stable-argument-order-5520-3

Author: Julian
Reviewer: namn, exarkun
Fixes: [#5520](https://github.com/twisted/twisted/issues/5520)

If given the names of multiple test suites on the command line, trial now runs them in the
same order as they appear on the command line.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_30"></a>|@exarkun commented|
|-|-|

I closed [#1948](https://github.com/twisted/twisted/issues/1948) as a duplicate, partly of this ticket and partly of [#5787](https://github.com/twisted/twisted/issues/5787).
